### PR TITLE
Add bun-types, add typechecking, add `child_process` types

### PR DESCRIPTION
**Pull request by colinhacks**
- Moves `bun-types` into `packages`
- Fixes types throughout the codebase. All code including tests (except some excluded subdirectories) are now typechecked according to the root `tsconfig.json`. There's no need to rebuild `bun-types` to get the latest types - the "source" type declaration files are used.
- Removes the labeling workflows from `bun-types`
- Add `child_process` types

You can run `bun run typecheck` to run typechecking. This is currently passing. We should consider integrating this into CI at some point if it isn't too burdensome.

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "tabWidth": 2,
+  "useTabs": false,
+  "singleQuote": false,
+  "bracketSpacing": true,
+  "trailingComma": "es5"
```
Comment (Jarred-Sumner) [inline .prettierrc]: can we do `"trailingComma": "all"`? 

it makes copy pasting code easier

File: examples/ssl.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import { resolve } from "path";
 
+Bun.file;
```
Comment (Jarred-Sumner) [inline examples/ssl.ts]: did you mean for this to be included

File: package.json [hunk 3]
```diff
@@ -1,23 +1,28 @@
 {
   "dependencies": {
-    "bun-types": "latest",
     "express": "^4.18.2",
     "mitata": "^0.1.3",
     "peechy": "0.4.32",
     "react": "next",
-    "react-dom": "next"
+    "react-dom": "next",
+    "svelte": "^3.52.0"
```
Comment (Jarred-Sumner) [inline package.json]: i don't remember why svelte is here

File: tsconfig.json [hunk 4]
```diff
@@ -1,11 +1,25 @@
...
+    "."
+  ],
+  "exclude": [
+    "src/test",
+    "packages",
+    "examples",
+    "bench",
+    "test/bun.js/solid-dom-fixtures",
+    "test/bun.js/bundled",
+    "src/bun.js/builtins",
```
Comment (Jarred-Sumner) [inline tsconfig.json:22]:
```suggestion
    // JavaScriptCore builtins use a non-standard "@" symbol to indicate a private identifier which no other tool supports
    "src/bun.js/builtins",
```

File: test/bun.js/websocket-server.test.ts [hunk 5]
```diff
@@ -604,7 +603,7 @@ describe("websocket server", () => {
           ws.send("hello world");
         },
         message(ws, msg) {
-          ws.cork(() => {
+          ws.cork!(() => {
```
Comment (Jarred-Sumner) [inline test/bun.js/websocket-server.test.ts]: is this missing from the types? `cork` is never undefined

File: test/bun.js/websocket-server.test.ts [hunk 6]
```diff
@@ -138,8 +138,7 @@ describe("websocket server", () => {
           if (
             server.upgrade(req, {
               data: "hello world",
-
-              headers: 1238,
+              headers: 1238 as any, // why is this a number?
```
Comment (Jarred-Sumner) [inline test/bun.js/websocket-server.test.ts]: we are testing that numbers/invalid types don't cause a crash

we will need a lot of tests like that tbh

File: src/api/demo/pages/scan.tsx [hunk 7]
```diff
@@ -1,19 +1,24 @@
 import Head from "next/head";
 import Image from "next/image";
+import { readFile } from "fs/promises";
 import styles from "../styles/Home.module.css";
 
+declare const globalThis: any;
```
Comment (Jarred-Sumner) [inline src/api/demo/pages/scan.tsx]: `globalThis` should always be defined?

File: tsconfig.base.json [hunk 8]
```diff
@@ -0,0 +1,24 @@
+{
+  "compilerOptions": {
+    "lib": [
+      "ESNext"
+    ],
+    "module": "esnext",
+    "target": "esnext",
+    "moduleResolution": "nodenext",
+    "strict": true,
```
Comment (Jarred-Sumner) [inline tsconfig.base.json:9]: I think `strict` for runtime tests will end up being a little bit of a velocity hit for us

File: test/bun.js/websocket-server.test.ts [hunk 9]
```diff
@@ -654,14 +653,14 @@ describe("websocket server", () => {
...
-            count: 0,
+            data: { count: 0 },
           })
         )
           return new Response("noooooo hello world");
       },
     });
 
-    await new Promise((resolve, reject) => {
+    await new Promise<void>((resolve, reject) => {
```
Comment (Jarred-Sumner) [inline test/bun.js/websocket-server.test.ts:663]: I don't think it's worth developer time for us to put `<void>` in here

Maybe we could enable noImplicitAny?

**Review by Jarred-Sumner — COMMENTED** [review]
Mostly looks good

some comments

File: package.json (hunk 3 above)
Comment (colinhacks) [reply package.json]: It's used by `bun-loader-svelte.ts` which is imported in `plugins.test.ts`. I think it makes sense for all the dependencies needed for tests to live in the root `package.json` but we could add a separate `package.json` in `tests/bun.js` if you prefer

File: packages/bun-types/bun.d.ts [hunk 10]
```diff
@@ -2729,6 +2729,97 @@ declare module "bun" {
   export const revision: string;
 }
 
+// Declare "static" methods in Error
+interface ErrorConstructor {
```
Comment (Jarred-Sumner) [inline packages/bun-types/bun.d.ts]: This should go in globals

File: test/bun.js/websocket-server.test.ts (hunk 9 above)
Comment (colinhacks) [reply test/bun.js/websocket-server.test.ts:663]: `noImplicitAny` isn't the problem here - this is necessary based on the typing of `Promise` in the `lib.esnext` typings. I don't think there's a setting in tsconfig that would make this valid without a type annotation.

Screenshot 2022-11-09 at 1 39 40 PM

File: tsconfig.base.json (hunk 8 above)
Comment (colinhacks) [reply tsconfig.base.json:9]: The changes required to modify existing tests compatible with strict mode were pretty minimal: https://github.com/oven-sh/bun/pull/1475/commits/737ef3c9f7879599f5faf9441238ea8a4b5c3d9f

Also part of the goal here is for the test suite to sanity check the typings as well, so this is necessary if we want to reflect a realistic user environment.

File: test/bun.js/websocket-server.test.ts (hunk 5 above)
Comment (colinhacks) [reply test/bun.js/websocket-server.test.ts]: Fixed

File: src/api/demo/pages/scan.tsx (hunk 7 above)
Comment (colinhacks) [reply src/api/demo/pages/scan.tsx]: Hm yep. For some reason this was giving me trouble at some point but seems to work fine without it now.

**Pull request merged**